internal/docstore/dynamodocstore/query.go

```diff
@@ -0,0 +1,170 @@
+		FilterExpression:          ce.Filter(),
+		ProjectionExpression:      ce.Projection(),
+		ExclusiveStartKey:         startAfter,
+	})
+	if err != nil {
+		return nil, err
+	}
+	return &documentIterator{
+		c:     c,
+		q:     q,
+		items: out.Items,
+		count: *out.Count,
+		limit: q.Limit, // manually count limit since dynamodb uses "limit" as scan limit before filtering
+		last:  out.LastEvaluatedKey,
+	}, nil
+}
+
+func processFilter(cb expression.Builder, fs []driver.Filter, pkey, skey string) expression.Builder {
+	// TODO(shantuo): process the partition key and remove the hard-coded key
+	// condition after we fix the API.
+	kbs := []expression.KeyConditionBuilder{expression.KeyEqual(expression.Key(pkey), expression.Value("query"))}
+	var cbs []expression.ConditionBuilder
+
+	for _, f := range fs {
+		if kb, ok := toKeyCondition(f, pkey, skey); ok {
+			kbs = append(kbs, kb)
+			continue
+		}
+		cbs = append(cbs, toFilter(f))
+	}
+	keyBuilder := kbs[0]
+	for i := 1; i < len(kbs); i++ {
+		keyBuilder.And(kbs[i])
+	}
+	cb = cb.WithKeyCondition(keyBuilder)
+	var condBuilder expression.ConditionBuilder
+	if len(cbs) > 0 {
+		condBuilder = cbs[0]
+		for i := 1; i < len(cbs); i++ {
+			condBuilder.And(cbs[i])
+		}
+		cb = cb.WithFilter(condBuilder)
+	}
+	return cb
+}
+
+func toKeyCondition(f driver.Filter, pkey, skey string) (expression.KeyConditionBuilder, bool) {
+	kp := strings.Join(f.FieldPath, ".")
+	if kp == skey {
+		key := expression.Key(kp)
+		val := expression.Value(f.Value)
+		switch f.Op {
+		case "<":
+			return expression.KeyLessThan(key, val), true
+		case "<=":
+			return expression.KeyLessThanEqual(key, val), true
+		case "=":
+			return expression.KeyEqual(key, val), true
+		case ">=":
+			return expression.KeyGreaterThanEqual(key, val), true
+		case ">":
+			return expression.KeyGreaterThan(key, val), true
+		}
+	}
+	return expression.KeyConditionBuilder{}, false
+}
+
+func toFilter(f driver.Filter) expression.ConditionBuilder {
+	name := expression.Name(strings.Join(f.FieldPath, "."))
+	val := expression.Value(f.Value)
+	switch f.Op {
+	case "<":
+		return expression.LessThan(name, val)
+	case "<=":
+		return expression.LessThanEqual(name, val)
+	case "=":
+		return expression.Equal(name, val)
+	case ">=":
+		return expression.GreaterThanEqual(name, val)
+	case ">":
+		return expression.GreaterThan(name, val)
+	}
+	return expression.ConditionBuilder{}
+}
+
+type documentIterator struct {
+	c     *collection
+	q     *driver.Query
+	items []map[string]*dynamodb.AttributeValue
+	count int64
+	curr  int
+	limit int
+	last  map[string]*dynamodb.AttributeValue
+}
+
+func (it *documentIterator) Next(ctx context.Context, doc driver.Document) error {
+	if it.limit > 0 && it.curr >= it.limit ||
+		it.curr >= len(it.items) && it.last == nil {
+		return io.EOF
+	}
+	if it.curr >= len(it.items) && it.last != nil {
+		// Make a new query request at the end of this page.
+		i, err := it.c.runGetQuery(ctx, it.q, it.last)
```

Comment:
It seems odd to me to create a new iterator and then copy from it. How about if runGetQuery returns the query output struct (what you call `out` above), and you use that to populate `it` here?

Reply:
Done.